File: tests/e2e/mobile-scroll-affordance.py
#!/usr/bin/env python3
"""Mobile scrollability + scroll-affordance a11y check (see repo docs)."""
import re



At 390x844 (mobile), verifies:
  1. /library main list is vertically scrollable and applies the shared
     touch tokens (touch-action: pan-y, overscroll-behavior-y: contain).
  2. A Radix Dialog (home "Ask OG Bot" trigger) opens, its [role="dialog"]
     content applies the same touch tokens, and has an accessible name.
  3. A Radix Sheet (mobile sidebar via the sidebar trigger) opens, its
     content applies the same touch tokens, and exposes role="dialog".
  4. Scroll-fade mask-image is gated by `prefers-contrast: no-preference`
     AND `prefers-reduced-transparency: no-preference`, so faded edge text
     is disabled for users who opted into higher contrast.

Exit 0 clean, 1 on any failure.

Run:
  python3 tests/e2e/mobile-scroll-affordance.py
"""
import asyncio
import json
import os
import sys
from playwright.async_api import async_playwright

BASE = os.environ.get("BASE_URL", "http://localhost:8080")
VIEWPORT = {"width": 390, "height": 844}

storage_key = os.environ.get("LOVABLE_BROWSER_SUPABASE_STORAGE_KEY")
session_json = os.environ.get("LOVABLE_BROWSER_SUPABASE_SESSION_JSON")
cookies_json = os.environ.get("LOVABLE_BROWSER_SUPABASE_COOKIES_JSON")

failures = []
def ok(msg):   print(f"  ✓ {msg}")
def bad(msg):  failures.append(msg); print(f"  ✗ {msg}")

async def restore_auth(context, page):
    if cookies_json:
        cookies = json.loads(cookies_json)
        for c in cookies:
            c["url"] = BASE
        await context.add_cookies(cookies)
    await page.goto(BASE, wait_until="domcontentloaded")
    if storage_key and session_json:
        await page.evaluate(
            f"window.localStorage.setItem({json.dumps(storage_key)}, {json.dumps(session_json)})"
        )

def assert_touch(info, label):
    ta = info.get("touchAction") or ""
    ob = info.get("overscrollBehaviorY") or ""
    (ok if "pan-y" in ta else bad)(f"{label}: touch-action includes pan-y (got '{ta}')")
    (ok if "contain" in ob else bad)(f"{label}: overscroll-behavior-y contain (got '{ob}')")

async def run():
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)

        # -------- 1) /library main list --------
        print("\n/library — main list scrollability")
        ctx = await browser.new_context(viewport=VIEWPORT)
        page = await ctx.new_page()
        await restore_auth(ctx, page)
        await page.goto(f"{BASE}/library", wait_until="domcontentloaded")
        try:
            await page.wait_for_selector('[data-testid="library-root"]', timeout=15000)
        except Exception as e:
            bad(f"library-root did not render: {e}")
        else:
            info = await page.evaluate("""() => {
                const el = document.querySelector('[data-testid=\"library-root\"]');
                const s = getComputedStyle(el);
                return {
                    touchAction: s.touchAction,
                    overscrollBehaviorY: s.overscrollBehaviorY,
                    dsh: document.documentElement.scrollHeight,
                    dch: document.documentElement.clientHeight,
                };
            }""")
            assert_touch(info, "library-root")
            if info["dsh"] <= info["dch"] + 4:
                ok("library fits without scroll (nothing to scroll)")
            else:
                scrolled = await page.evaluate("""() => {
                    const targets = [...document.querySelectorAll('main,[data-scroll-container],.overflow-y-auto,.overflow-auto')];
                    for (const t of targets) {
                        const before = t.scrollTop; t.scrollTop = 400;
                        if (t.scrollTop > before) return { via: 'container', delta: t.scrollTop - before };
                    }
                    const before = window.scrollY; window.scrollTo(0, 400);
                    return { via: 'window', delta: window.scrollY - before };
                }""")
                if scrolled["delta"] > 10:
                    ok(f"library scrolled via {scrolled['via']} (+{scrolled['delta']}px)")
                else:
                    bad(f"library did not scroll (via={scrolled['via']} delta={scrolled['delta']})")
        await ctx.close()

        # -------- 2) Radix Dialog on / --------
        print("\n/ — Radix Dialog (Ask OG Bot)")
        ctx = await browser.new_context(viewport=VIEWPORT)
        page = await ctx.new_page()
        await restore_auth(ctx, page)
        await page.goto(f"{BASE}/", wait_until="domcontentloaded")
        await page.wait_for_timeout(1200)
        try:
            trigger = page.get_by_role("button", name=lambda n: n and ("Start any task" in n or "Ask OG Bot" in n)).first
            await trigger.wait_for(timeout=10000)
            await trigger.click()
            dialog = page.locator('[role="dialog"]').first
            await dialog.wait_for(timeout=5000)
            info = await dialog.evaluate("""el => {
                const s = getComputedStyle(el);
                return {
                    touchAction: s.touchAction,
                    overscrollBehaviorY: s.overscrollBehaviorY,
                    role: el.getAttribute('role'),
                    hasLabel: !!(el.getAttribute('aria-labelledby') || el.getAttribute('aria-label')),
                };
            }""")
            assert_touch(info, "dialog")
            (ok if info["role"] == "dialog" else bad)(f"dialog role=dialog (got '{info['role']}')")
            (ok if info["hasLabel"] else bad)("dialog has accessible name")
        except Exception as e:
            bad(f"dialog check failed: {e}")
        await ctx.close()

        # -------- 3) Radix Sheet (mobile sidebar) --------
        print("\n/ — Radix Sheet (mobile sidebar)")
        ctx = await browser.new_context(viewport=VIEWPORT)
        page = await ctx.new_page()
        await restore_auth(ctx, page)
        await page.goto(f"{BASE}/", wait_until="domcontentloaded")
        await page.wait_for_timeout(1000)
        try:
            trigger = page.locator('[data-sidebar="trigger"]').first
            await trigger.wait_for(timeout=10000)
            await trigger.click()
            sheet = page.locator('[data-sidebar="sidebar"][data-mobile="true"], [role="dialog"]').first
            await sheet.wait_for(timeout=5000)
            info = await sheet.evaluate("""el => {
                const s = getComputedStyle(el);
                return {
                    touchAction: s.touchAction,
                    overscrollBehaviorY: s.overscrollBehaviorY,
                    role: el.getAttribute('role'),
                    hasLabel: !!(el.getAttribute('aria-labelledby') || el.getAttribute('aria-label') || el.querySelector('h1,h2,h3')),
                };
            }""")
            assert_touch(info, "sheet")
            (ok if info["role"] == "dialog" else bad)(f"sheet role=dialog (got '{info['role']}')")
            (ok if info["hasLabel"] else bad)("sheet has accessible name")
        except Exception as e:
            bad(f"sheet check failed: {e}")
        await ctx.close()

        # -------- 4) Scroll-fade a11y gating --------
        print("\nscroll-fade mask gated by prefers-contrast / prefers-reduced-transparency")
        ctx = await browser.new_context(viewport=VIEWPORT)
        page = await ctx.new_page()
        await restore_auth(ctx, page)
        await page.goto(f"{BASE}/library", wait_until="domcontentloaded")
        await page.wait_for_timeout(500)
        guarded = await page.evaluate("""() => {
            const wanted = ['prefers-contrast', 'prefers-reduced-transparency'];
            for (const sheet of Array.from(document.styleSheets)) {
                try {
                    for (const rule of Array.from(sheet.cssRules || [])) {
                        const t = rule.cssText || '';
                        if (t.includes('mask-image') && wanted.every(w => t.includes(w))) return true;
                    }
                } catch (e) {}
            }
            return false;
        }""")
        (ok if guarded else bad)("mask-image is gated behind both a11y prefs media queries")
        await ctx.close()

        await browser.close()

    if failures:
        print(f"\n{len(failures)} failure(s):", file=sys.stderr)
        for f in failures: print("  - " + f, file=sys.stderr)
        sys.exit(1)
    print("\nAll mobile scroll + a11y checks passed.")

asyncio.run(run())

File: scripts/mobile-scroll-affordance.mjs
#!/usr/bin/env node
/**
 * Mobile scrollability + scroll-affordance a11y check.
 *
 * At 390x844 (mobile), verifies:
 *   1. /library main list is vertically scrollable (scrollTop actually
 *      advances when we scroll) and applies the shared touch tokens
 *      (touch-action: pan-y, overscroll-behavior-y: contain).
 *   2. A Radix Dialog (home page "Ask OG Bot" trigger) opens, its
 *      [role="dialog"] content applies the same touch tokens, and
 *      keeps an accessible name (aria-labelledby or aria-label).
 *   3. A Radix Sheet (mobile sidebar via the sidebar trigger) opens,
 *      its content applies the same touch tokens, and exposes a
 *      role="dialog" node to assistive tech.
 *   4. Scroll-fade mask is disabled under `prefers-contrast: more`
 *      and `prefers-reduced-transparency: reduce`, so faded edge text
 *      never drops below the AA contrast the design tokens guarantee.
 *
 * Exits 0 clean, 1 on any failure.
 */
import { chromium } from "playwright";

const BASE = process.env.BASE_URL ?? "http://localhost:8080";
const VIEWPORT = { width: 390, height: 844 };

const storageKey = process.env.LOVABLE_BROWSER_SUPABASE_STORAGE_KEY;
const sessionJson = process.env.LOVABLE_BROWSER_SUPABASE_SESSION_JSON;
const cookiesJson = process.env.LOVABLE_BROWSER_SUPABASE_COOKIES_JSON;

const failures = [];
const pass = (m) => console.log(`  ✓ ${m}`);
const fail = (m) => { failures.push(m); console.log(`  ✗ ${m}`); };

async function restoreAuth(context, page) {
  if (cookiesJson) {
    const cookies = JSON.parse(cookiesJson).map((c) => ({ ...c, url: BASE }));
    await context.addCookies(cookies);
  }
  await page.goto(BASE, { waitUntil: "domcontentloaded" });
  if (storageKey && sessionJson) {
    await page.evaluate(
      ([k, v]) => window.localStorage.setItem(k, v),
      [storageKey, sessionJson],
    );
  }
}

function expectTouchTokens(styles, label) {
  if (!/pan-y/.test(styles.touchAction || "")) {
    fail(`${label}: touch-action should include pan-y (got "${styles.touchAction}")`);
  } else pass(`${label}: touch-action pan-y`);
  if (!/contain/.test(styles.overscrollBehaviorY || "")) {
    fail(`${label}: overscroll-behavior-y should be contain (got "${styles.overscrollBehaviorY}")`);
  } else pass(`${label}: overscroll-behavior-y contain`);
}

async function run() {
  const browser = await chromium.launch({ headless: true });

  // ---------- 1) /library main list scrollable + touch tokens ----------
  console.log("\n/library — main list scrollability");
  {
    const ctx = await browser.newContext({ viewport: VIEWPORT });
    const page = await ctx.newPage();
    await restoreAuth(ctx, page);
    await page.goto(`${BASE}/library`, { waitUntil: "domcontentloaded" });
    await page.waitForSelector('[data-testid="library-root"]', { timeout: 15000 });

    const info = await page.evaluate(() => {
      const el = document.querySelector('[data-testid="library-root"]');
      const s = getComputedStyle(el);
      return {
        touchAction: s.touchAction,
        overscrollBehaviorY: s.overscrollBehaviorY,
        docScrollHeight: document.documentElement.scrollHeight,
        docClientHeight: document.documentElement.clientHeight,
      };
    });
    expectTouchTokens(info, "library-root");

    if (info.docScrollHeight <= info.docClientHeight + 4) {
      pass("library page fits without scroll (nothing to assert)");
    } else {
      // Try main-shell scroll first, then window as fallback.
      const scrolled = await page.evaluate(() => {
        const targets = [
          ...document.querySelectorAll('main, [data-scroll-container], .overflow-y-auto, .overflow-auto'),
        ];
        for (const t of targets) {
          const before = t.scrollTop;
          t.scrollTop = 400;
          if (t.scrollTop > before) return { via: "container", delta: t.scrollTop - before };
        }
        const before = window.scrollY;
        window.scrollTo(0, 400);
        return { via: "window", delta: window.scrollY - before };
      });
      if (scrolled.delta > 10) pass(`library scrolled via ${scrolled.via} (+${scrolled.delta}px)`);
      else fail(`library did not scroll (via=${scrolled.via}, delta=${scrolled.delta})`);
    }
    await ctx.close();
  }

  // ---------- 2) Radix Dialog on home ("Ask OG Bot") ----------
  console.log("\n/ — Radix Dialog (Ask OG Bot)");
  {
    const ctx = await browser.newContext({ viewport: VIEWPORT });
    const page = await ctx.newPage();
    await restoreAuth(ctx, page);
    await page.goto(`${BASE}/`, { waitUntil: "domcontentloaded" });
    await page.waitForTimeout(1200);

    const trigger = page.getByRole("button", { name: /start any task|ask og bot/i }).first();
    await trigger.waitFor({ timeout: 10000 });
    await trigger.click();

    const dialog = page.locator('[role="dialog"]').first();
    await dialog.waitFor({ timeout: 5000 });

    const info = await dialog.evaluate((el) => {
      const s = getComputedStyle(el);
      return {
        touchAction: s.touchAction,
        overscrollBehaviorY: s.overscrollBehaviorY,
        hasLabel:
          !!el.getAttribute("aria-labelledby") ||
          !!el.getAttribute("aria-label"),
        role: el.getAttribute("role"),
      };
    });
    expectTouchTokens(info, "dialog");
    if (info.role === "dialog") pass("dialog exposes role=dialog");
    else fail(`dialog role missing (got "${info.role}")`);
    if (info.hasLabel) pass("dialog has accessible name (aria-labelledby/label)");
    else fail("dialog missing accessible name");

    await ctx.close();
  }

  // ---------- 3) Radix Sheet (mobile sidebar) ----------
  console.log("\n/ — Radix Sheet (mobile sidebar)");
  {
    const ctx = await browser.newContext({ viewport: VIEWPORT });
    const page = await ctx.newPage();
    await restoreAuth(ctx, page);
    await page.goto(`${BASE}/`, { waitUntil: "domcontentloaded" });
    await page.waitForTimeout(1000);

    const trigger = page.locator('[data-sidebar="trigger"]').first();
    await trigger.waitFor({ timeout: 10000 });
    await trigger.click();

    const sheet = page.locator('[data-sidebar="sidebar"][data-mobile="true"], [role="dialog"]').first();
    await sheet.waitFor({ timeout: 5000 });

    const info = await sheet.evaluate((el) => {
      const s = getComputedStyle(el);
      return {
        touchAction: s.touchAction,
        overscrollBehaviorY: s.overscrollBehaviorY,
        role: el.getAttribute("role"),
        hasLabel:
          !!el.getAttribute("aria-labelledby") ||
          !!el.getAttribute("aria-label") ||
          !!el.querySelector("h1,h2,h3,[data-slot='sheet-title']"),
      };
    });
    expectTouchTokens(info, "sheet");
    if (info.role === "dialog") pass("sheet exposes role=dialog");
    else fail(`sheet role missing (got "${info.role}")`);
    if (info.hasLabel) pass("sheet has accessible name");
    else fail("sheet missing accessible name");

    await ctx.close();
  }

  // ---------- 4) Scroll-fade disabled under a11y prefs ----------
  console.log("\nscroll-fade mask respects prefers-contrast / prefers-reduced-transparency");
  for (const prefs of [
    [{ name: "prefers-contrast", value: "more" }],
    [{ name: "prefers-reduced-transparency", value: "reduce" }],
  ]) {
    const ctx = await browser.newContext({ viewport: VIEWPORT });
    await ctx.emulateMedia({ reducedMotion: "no-preference" });
    const page = await ctx.newPage();
    await page.emulateMedia({ media: "screen", forcedColors: "none" });
    // Playwright doesn't yet expose prefers-contrast directly; force via CSSOM.
    await page.addInitScript((prefsList) => {
      const orig = window.matchMedia.bind(window);
      window.matchMedia = (q) => {
        for (const p of prefsList) {
          if (q.includes(p.name) && q.includes(p.value)) {
            return { matches: true, media: q, addListener() {}, removeListener() {}, addEventListener() {}, removeEventListener() {}, dispatchEvent() { return false; }, onchange: null };
          }
        }
        return orig(q);
      };
    }, prefs);
    await restoreAuth(ctx, page);
    await page.goto(`${BASE}/library`, { waitUntil: "domcontentloaded" });
    await page.waitForSelector('[data-testid="library-root"]', { timeout: 15000 });
    // Note: matchMedia patch doesn't feed CSS media queries — CSS is evaluated
    // by the engine, not JS. So we assert the *rule* exists in the stylesheet.
    const guarded = await page.evaluate(() => {
      const wanted = ["prefers-contrast", "prefers-reduced-transparency"];
      for (const sheet of Array.from(document.styleSheets)) {
        try {
          for (const rule of Array.from(sheet.cssRules || [])) {
            if (rule.cssText && wanted.every((w) => rule.cssText.includes(w))) return true;
          }
        } catch {/* cross-origin */ }
      }
      return false;
    });
    const label = prefs[0].name;
    if (guarded) pass(`mask-image guarded by ${label} media query`);
    else fail(`mask-image is NOT gated by ${label} media query`);
    await ctx.close();
    break; // one iteration is enough; both prefs are on the same rule.
  }

  await browser.close();
  if (failures.length) {
    console.error(`\n${failures.length} failure(s):`);
    for (const f of failures) console.error("  - " + f);
    process.exit(1);
  }
  console.log("\nAll mobile scroll + a11y checks passed.");
}

run().catch((e) => { console.error(e); process.exit(1); });
